src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,146 @@
+
+                            for (var i = 0; i < methodParameters.Length; i++)
+                            {
+                                // If the method we're calling accepts a cancellation token, we wouldn't have passed that, so add in a CancellationToken.None here. Although we could
+                                // remove the cancellation from the underlying method, this keeps that support around should we need it.
+                                if (i == methodParameters.Length - 1 && methodParameters[i].ParameterType == typeof(CancellationToken))
+                                    arguments[i] = CancellationToken.None;
+                                else
+                                    arguments[i] = request.Parameters[i].ToObject(methodParameters[i].ParameterType);
+                            }
+
+                            var result = method.Invoke(rpcTarget, arguments);
+
+                            if (result is Task task)
+                            {
+                                await task.ConfigureAwait(false);
+
+                                // If it's actually a Task<T> then get the result; we're looking at the declared return type because in some cases a method might
+                                // return a Task<T> under the covers as a workaround for the lack of TaskCompletionSource on .NET Framework but we don't want to see
+                                // that workaround since the result isn't intended to be seen.
+                                if (method.ReturnType.IsConstructedGenericType)
+                                {
+                                    result = task.GetType().GetProperty("Result")!.GetValue(task);
+                                }
+                                else
+                                {
+                                    // It's just a simple Task so no result to return
+                                    result = null;
+                                }
+                            }
+
+                            response = new Response { Id = request.Id, Value = result is not null ? JToken.FromObject(result) : null };
+                        }
+                        catch (Exception e)
+                        {
+                            if (e is TargetInvocationException)
+                                e = e.InnerException ?? e;
+
+                            response = new Response { Id = request.Id, Exception = $"An exception of type {e.GetType()} was thrown: {e.Message}" };
+                        }
+
+                        var responseJson = JsonConvert.SerializeObject(response, JsonSettings.SingleLineSerializerSettings);
+
+#if DEBUG
+                        // Assert we didn't put a newline in this, since if we did the receiving side won't know how to parse it
+                        Contract.ThrowIfTrue(responseJson.Contains("\r") || responseJson.Contains("\n"));
```

Comment:
This was being covered by tests, otherwise it means we're running this over not-small strings in the case of the larger responses. If you want the extra check sure, it's just extra CPU in production which isn't exactly cheap here!